src/set-core.js

```diff
@@ -2,10 +2,37 @@ var h = require("./helpers");
 var clause = require("./clause");
 var compare = require("./compare");
 var get = require("./get");
-var assign = require("can-util/js/assign/assign");
+var assign = require("can-assign");
 var each = require("can-util/js/each/each");
 var makeArray = require("can-util/js/make-array/make-array");
 var isEmptyObject = require("can-util/js/is-empty-object/is-empty-object");
+var getProps = require("can-util/js/get/get");
+
+// concatUnique
+// concat all items in bItems onto aItems that do not already exist in aItems.
+// same-object and ID collisions are both looked at when deciding whether
+// an item matches another.
+function concatUnique(aItems, bItems, algebra) {
+	var keyTree = {};
+
+	aItems.forEach(function(item) {
+		var keyNode = keyTree;
+		each(algebra.clauses.id, function(prop) {
+			keyNode = keyNode[getProps(item, prop)] = keyNode[getProps(item, prop)] || {};
+		});
+	});
+
+	return aItems.concat(bItems.filter(function(item) {
+		var keyNode = keyTree;
+		if(aItems.indexOf(item) > -1) {
```

Comment:
this makes this concat `O(n^2)`.  Use of `Set` would be nice here ... or drop this b/c `id` "should" take care of it.

Reply:
Set (CIDSet for compatibility) is a good idea.  As mentioned above, if there is no `id` clause, testing for same object will catch duplicates in some use cases.